tket2/src/extension/angle.rs

```diff
@@ -136,49 +126,51 @@ impl CustomConst for ConstAngle {
         hugr::values::downcast_equal_consts(self, other)
     }
 }
-/// Collect a vector into an array.
-fn collect_array<const N: usize, T: std::fmt::Debug>(arr: &[T]) -> [&T; N] {
-    arr.iter().collect_vec().try_into().unwrap()
+
+fn type_var(var_id: usize, extension: &Extension) -> Result<Type, SignatureError> {
+    Ok(Type::new_extension(angle_def(extension).instantiate(
+        vec![TypeArg::new_var_use(var_id, LOG_DENOM_TYPE_PARAM)],
+    )?))
 }
+fn atrunc_sig(extension: &Extension) -> Result<FunctionType, SignatureError> {
+    let in_angle = type_var(0, extension)?;
+    let out_angle = type_var(1, extension)?;
 
-fn atrunc_sig(arg_values: &[TypeArg]) -> Result<FunctionType, SignatureError> {
-    let [arg0, arg1] = collect_array(arg_values);
-    let m: u8 = get_log_denom(arg0)?;
-    let n: u8 = get_log_denom(arg1)?;
-    if m < n {
-        return Err(SignatureError::InvalidTypeArgs);
-    }
-    Ok(FunctionType::new(
-        vec![angle_type(arg0.clone())],
-        vec![angle_type(arg1.clone())],
-    ))
+    Ok(FunctionType::new(vec![in_angle], vec![out_angle]))
 }
 
-fn aconvert_sig(arg_values: &[TypeArg]) -> Result<FunctionType, SignatureError> {
-    let [arg0, arg1] = collect_array(arg_values);
+fn aconvert_sig(extension: &Extension) -> Result<FunctionType, SignatureError> {
+    let in_angle = type_var(0, extension)?;
+    let out_angle = type_var(1, extension)?;
     Ok(FunctionType::new(
-        vec![angle_type(arg0.clone())],
-        vec![Type::new_sum(vec![angle_type(arg1.clone()), ERROR_TYPE])],
+        vec![in_angle],
+        vec![Type::new_sum(vec![out_angle, ERROR_TYPE])],
     ))
 }
 
+/// Collect a vector into an array.
+fn collect_array<const N: usize, T: std::fmt::Debug>(arr: &[T]) -> [&T; N] {
+    arr.iter().collect_vec().try_into().unwrap()
+}
+
 fn abinop_sig(arg_values: &[TypeArg]) -> Result<FunctionType, SignatureError> {
     let [arg0, arg1] = collect_array(arg_values);
     let m: u8 = get_log_denom(arg0)?;
     let n: u8 = get_log_denom(arg1)?;
     let l: u8 = max(m, n);
     Ok(FunctionType::new(
-        vec![
-            angle_type(TypeArg::BoundedNat { n: m as u64 }),
-            angle_type(TypeArg::BoundedNat { n: n as u64 }),
-        ],
-        vec![angle_type(TypeArg::BoundedNat { n: l as u64 })],
+        vec![angle_type(m), angle_type(n)],
+        vec![angle_type(l)],
     ))
 }
 
-fn aunop_sig(arg_values: &[TypeArg]) -> Result<FunctionType, SignatureError> {
-    let [arg] = collect_array(arg_values);
-    Ok(FunctionType::new_linear(vec![angle_type(arg.clone())]))
+fn aunop_sig(extension: &Extension) -> Result<FunctionType, SignatureError> {
+    let angle = type_var(0, extension)?;
+    Ok(FunctionType::new_linear(vec![angle]))
+}
+
+fn angle_def(extension: &Extension) -> &TypeDef {
```

Comment:
Maybe this is a wider question about the extensions API, but it seems strange that this function (and some others -- `type_var`, `angle_custom_type`, `atrunc_sig` etc) need to take their own extension as an argument. I suppose it is because `angle_custom_type` is exposed at the module level? But it feels like it should be unnecessary.

Reply:
OK, maybe something to ponder but this LGTM.

Reply:
I have an issue to raise already about needing a temporary register containing types in order to be able to create validated type schemes, perhaps we can rethink the definition API as part of that.